src/programs/bamfiltermc.1
.TH BAMFILTERMC 1 "July 2016" BIOBAMBAM2
.SH NAME
bamfiltermc - remove MC aux fields for reads with unmapped mates
.SH SYNOPSIS
.PP
.B bamfilterflags
[options]
.SH DESCRIPTION
bamfiltermc reads a SAM/BAM/CRAM file from standard input, 
removes MC aux fields for reads with unmapped mates (which were erroneously
inserted by earlier versions of biobambam2)
and writes the resulting alignments to a SAM/BAM/CRAM file on the standard output stream.
.PP
The following key=value pairs can be given:
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output BAM file. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B numthreads=<1>: 
number of worker threads.
.PP
.B inputthreads=<1>: 
number of input threads.
.PP
.B outputthreads=<1>: 
number of output threads.
.PP
.B tmpfile=<filename>: 
prefix for temporary files. By default the temporary files are created in the current directory
.PP
.B md5=<0|1>:
md5 checksum creation for output file. Valid values are
.IP 0:
do not compute checksum. This is the default.
.IP 1:
compute checksum. If the md5filename key is set, then the checksum is
written to the given file. If md5filename is unset, then no checksum will be computed.
.PP
.B md5filename
file name for md5 checksum if md5=1.
.PP
.B index=<0|1>:
compute BAM index for output file. Valid values are
.IP 0:
do not compute BAM index. This is the default.
.IP 1:
compute BAM index. If the indexfilename key is set, then the BAM index is
written to the given file. If indexfilename is unset, then no BAM index will be computed.
.PP
.B indexfilename
file name for BAM index if index=1.
.PP
.B outputformat=<bam>: output file format.
All versions of bamsort come with support for the BAM output format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit). This format is not advisable for data sorted by query name.
.PP
.B inputformat=<bam>: input file format.
All versions of bamsort come with support for the BAM input format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit)
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2016 German Tischler, \(co 2011-2013 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.52) unstable; urgency=medium

  * add bamfiltermc program

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 12:39:39 +0200

biobambam2 (2.0.51) unstable; urgency=medium

  * only copy mate cigar string if mate is mapped in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 10:43:26 +0200

biobambam2 (2.0.50) unstable; urgency=medium

  * do not copy CIGAR data of unmapped mates in bamsormadup 

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jul 2016 16:13:07 +0200

biobambam2 (2.0.49) unstable; urgency=medium

  * update to libmaus2 version 2.0.216 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 29 Jun 2016 10:38:45 +0200

biobambam2 (2.0.48) unstable; urgency=medium

  * use AdapterFilter object as const in bamadapterfind
  * add blocksortverbose parameter in bamsormadup
  * reflect API update in libmaus2

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 28 Jun 2016 12:28:37 +0200

biobambam2 (2.0.47) unstable; urgency=medium

  * filter out alignments extending past the end of a reference sequence in filtersam

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Jun 2016 12:22:09 +0200

biobambam2 (2.0.46) unstable; urgency=medium

  * add hash sort option in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 12 Jun 2016 14:57:04 +0200

biobambam2 (2.0.45) unstable; urgency=medium

  * do not fail for alignments with no read base data in bamheap2
  * add options for turning N bases randomly to A,C,G,T and to convert bases to upper case in normalisefasta
  * Added a PG line for bam12auxmerge. (by Andrew Whitwham)

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 10 Jun 2016 11:08:55 +0200

biobambam2 (2.0.44) unstable; urgency=medium

  * reduce input memory per block in bamsormadup if space required is high

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 17 May 2016 10:36:56 +0200

biobambam2 (2.0.43) unstable; urgency=medium

  * improve diagnosis of low memory situations (autoarraytrace option in libmaus2)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 10 May 2016 09:10:42 +0200

biobambam2 (2.0.42) unstable; urgency=medium

  * update to libmaus2 2.0.179 and fix EOF detection bug on pipes 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 02 May 2016 11:03:36 +0200

biobambam2 (2.0.41) unstable; urgency=medium

  * update to libmaus2 2.0.178
  * add number of alignments kept in verbose output of bamfilternames

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 02 May 2016 10:33:25 +0200

biobambam2 (2.0.40) unstable; urgency=medium

  * print memory usage at end of bamcat in verbose mode

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 25 Apr 2016 11:23:05 +0200

biobambam2 (2.0.39) unstable; urgency=medium

  * fix collation in bamtofastq etc concerning comparison of reads names which only differ by leading zeros in number comparison 

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 19 Apr 2016 11:25:57 +0200

biobambam2 (2.0.38) unstable; urgency=medium

  * fix for bamrecalculatecigar concerning CIGAR strings starting with a deletion

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Apr 2016 09:27:23 +0200

biobambam2 (2.0.37) unstable; urgency=medium

  * zlib interface changes in libmaus2 for supporting alternative zlibs at run time 

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 07 Apr 2016 13:40:07 +0200

biobambam2 (2.0.36) unstable; urgency=medium

  * add name hash based sub sampling in bamdownsamplerandom

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Apr 2016 08:12:57 +0200

biobambam2 (2.0.35) unstable; urgency=medium

  * use factory class in ranking output for bamtofastq
  * use factory class in bamtoname
  * cleanup bamdownsamplerandom source and use factory class to instantiate alignment file decoder

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 30 Mar 2016 11:12:32 +0200

biobambam2 (2.0.34) unstable; urgency=medium

  * deduce index and md5 filenames from output file name if possible

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 29 Mar 2016 10:19:51 +0200

biobambam2 (2.0.33) unstable; urgency=medium

  * update manual page for bamsort 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 16 Mar 2016 21:30:17 +0100

biobambam2 (2.0.32) unstable; urgency=medium

  * add rmdup option in bamsort 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 16 Mar 2016 21:24:31 +0100

biobambam2 (2.0.31) unstable; urgency=medium

  * fix setting sort order field in bamsort/bamsormadup in case there is no sort order field in the input file 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 07 Mar 2016 15:05:13 +0100

biobambam2 (2.0.30) unstable; urgency=medium

  * do not exit/fail bam indexing when encountering reads with mapping coordinates but flagged as unmapped as sole reads assigned to a reference sequence

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 17 Feb 2016 10:05:51 +0100

biobambam2 (2.0.29) unstable; urgency=medium

  * add murmur3 based digest option in bamseqchksum 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 11 Feb 2016 16:10:33 +0100

biobambam2 (2.0.28) unstable; urgency=medium

  * add bamrecalculatecigar program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 10 Feb 2016 14:14:04 +0100

biobambam2 (2.0.27) unstable; urgency=medium

  * add fastaexplode program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 20 Jan 2016 15:27:26 +0100

biobambam2 (2.0.26) unstable; urgency=medium

  * update to libmaus2 2.0.114
  * Set variable to identify programs to iRODS. (by Andrew Whitwham)

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 12 Jan 2016 19:53:21 +0100

biobambam2 (2.0.25) unstable; urgency=medium

  * update to libmaus2 2.0.102 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 05 Nov 2015 15:28:29 +0100

biobambam2 (2.0.24) unstable; urgency=medium

  * update to libmaus2 2.0.99 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 18:33:56 +0100

biobambam2 (2.0.23) unstable; urgency=medium

  * update to libmaus2 2.0.98 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 10:49:14 +0100


src/programs/bamfiltermc.cpp
/**
    biobambam2
    Copyright (C) 2009-2016 German Tischler
    Copyright (C) 2011-2014 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include <libmaus2/aio/OutputStreamInstance.hpp>
#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/bambam/BamMultiAlignmentDecoderFactory.hpp>
#include <libmaus2/fastx/FastAIndex.hpp>
#include <libmaus2/math/IntegerInterval.hpp>
#include <libmaus2/math/iabs.hpp>
#include <libmaus2/util/GrowingFreeList.hpp>
#include <libmaus2/util/Histogram.hpp>
#include <libmaus2/util/TempFileRemovalContainer.hpp>
#include <libmaus2/bambam/BamHeaderUpdate.hpp>
#include <libmaus2/bambam/BgzfDeflateOutputCallbackBamIndex.hpp>
#include <libmaus2/lz/BgzfDeflateOutputCallbackMD5.hpp>
#include <libmaus2/bambam/StrCmpNum.hpp>
#include <libmaus2/util/FreeList.hpp>
#include <libmaus2/util/SimpleQueue.hpp>

#include <biobambam2/BamBamConfig.hpp>
#include <biobambam2/Licensing.hpp>

#include <config.h>

static int getDefaultMD5() { return 0; }
static int getDefaultIndex() { return 0; }

static int getDefaultVerbose()
{
	return 1;
}

static int getDefaultDisableValidation()
{
	return false;
}

static std::string getDefaultInputFormat()
{
	return "bam";
}

static int getDefaultNumThreads()
{
	return 1;
}


struct BamAlignmentFreeListDefaultTypeInfo
{
	typedef libmaus2::bambam::BamAlignment element_type;
	typedef element_type::shared_ptr_type pointer_type;

	static pointer_type deallocate(pointer_type /* p */)
	{
		return getNullPointer();
	}

	static pointer_type getNullPointer()
	{
		pointer_type p;
		return p;
	}
};

struct BamAlignmentFreeListDefaultAllocator
{
	typedef libmaus2::bambam::BamAlignment element_type;

	BamAlignmentFreeListDefaultAllocator()
	{
	}
	~BamAlignmentFreeListDefaultAllocator()
	{
	}

	element_type::shared_ptr_type operator()() const
	{
		element_type::shared_ptr_type ptr(new element_type);
		return ptr;
	}
};

void handleQueue(
	libmaus2::util::SimpleQueue < libmaus2::bambam::BamAlignment::shared_ptr_type > & Q,
	libmaus2::util::FreeList < libmaus2::bambam::BamAlignment, BamAlignmentFreeListDefaultAllocator, BamAlignmentFreeListDefaultTypeInfo > & FL,
	libmaus2::bambam::BamBlockWriterBase & wr,
	libmaus2::bambam::BamAuxFilterVector const & auxvec,
	uint64_t const numthreads
)
{
	#if defined(_OPENMP)
	#pragma omp parallel for num_threads(numthreads)
	#endif
	for ( uint64_t i = 0; i < Q.size(); ++i )
	{
		libmaus2::bambam::BamAlignment::shared_ptr_type P = Q[i];

		if ( P->isMateUnmap() )
			P->filterOutAux(auxvec);
	}

	while ( ! Q.empty() )
	{
		libmaus2::bambam::BamAlignment::shared_ptr_type P = Q.pop_front();
		wr.writeAlignment(*P);
		FL.put(P);
	}
}

int bamfiltermc(libmaus2::util::ArgInfo const & arginfo)
{
	bool const verbose = arginfo.getValue("verbose",getDefaultVerbose());

	libmaus2::bambam::BamAlignmentDecoderWrapper::unique_ptr_type decwrapper(
		libmaus2::bambam::BamMultiAlignmentDecoderFactory::construct(arginfo));
	::libmaus2::bambam::BamAlignmentDecoder * ppdec = &(decwrapper->getDecoder());
	::libmaus2::bambam::BamAlignmentDecoder & dec = *ppdec;
	::libmaus2::bambam::BamHeader const & header = dec.getHeader();
	::libmaus2::bambam::BamAlignment & algn = dec.getAlignment();
	std::string const tmpfilenamebase = arginfo.getUnparsedValue("tmpfile",arginfo.getDefaultTmpFileName());
	uint64_t const numthreads = arginfo.getValueUnsignedNumeric<uint64_t>("numthreads",getDefaultNumThreads());

	/*
	 * start index/md5 callbacks
	 */
	std::string const tmpfileindex = tmpfilenamebase + "_index";
	::libmaus2::util::TempFileRemovalContainer::addTempFile(tmpfileindex);

	std::string md5filename;
	std::string indexfilename;

	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > cbs;
	::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Pmd5cb;
	if ( arginfo.getValue<unsigned int>("md5",getDefaultMD5()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo) != std::string() )
			md5filename = libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo);
		else
			std::cerr << "[V] no filename for md5 given, not creating hash" << std::endl;

		if ( md5filename.size() )
		{
			::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Tmd5cb(new ::libmaus2::lz::BgzfDeflateOutputCallbackMD5);
			Pmd5cb = UNIQUE_PTR_MOVE(Tmd5cb);
			cbs.push_back(Pmd5cb.get());
		}
	}
	libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Pindex;
	if ( arginfo.getValue<unsigned int>("index",getDefaultIndex()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo) != std::string() )
			indexfilename = libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo);
		else
			std::cerr << "[V] no filename for index given, not creating index" << std::endl;

		if ( indexfilename.size() )
		{
			libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Tindex(new libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex(tmpfileindex));
			Pindex = UNIQUE_PTR_MOVE(Tindex);
			cbs.push_back(Pindex.get());
		}
	}
	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > * Pcbs = 0;
	if ( cbs.size() )
		Pcbs = &cbs;
	/*
	 * end md5/index callbacks
	 */

	::libmaus2::bambam::BamHeader::unique_ptr_type genuphead(
		libmaus2::bambam::BamHeaderUpdate::updateHeader(arginfo,header,"bamfiltermc",std::string(PACKAGE_VERSION))
	);
	libmaus2::bambam::BamBlockWriterBase::unique_ptr_type Pwriter(libmaus2::bambam::BamBlockWriterBaseFactory::construct(*genuphead,arginfo,Pcbs));
	libmaus2::bambam::BamBlockWriterBase & wr = *Pwriter;

	// freelist size
	uint64_t const flsize = 16*1024;
	libmaus2::util::FreeList < libmaus2::bambam::BamAlignment, BamAlignmentFreeListDefaultAllocator, BamAlignmentFreeListDefaultTypeInfo > FL(flsize);
	libmaus2::util::SimpleQueue < libmaus2::bambam::BamAlignment::shared_ptr_type > Q;

	libmaus2::bambam::BamAuxFilterVector auxvec;
	auxvec.set('M','C');

	uint64_t alcnt = 0;
	while ( dec.readAlignment() )
	{
		if ( FL.empty() )
			handleQueue(Q,FL,wr,auxvec,numthreads);
		assert ( ! FL.empty() );

		libmaus2::bambam::BamAlignment::shared_ptr_type P = FL.get();
		P->swap(algn);

		Q.push_back(P);

		if ( verbose && ((++alcnt % (1024*1024)) == 0) )
			std::cerr << "[V] " << alcnt << std::endl;
	}

	handleQueue(Q,FL,wr,auxvec,numthreads);

	// reset BAM writer
	Pwriter.reset();

	if ( Pmd5cb )
		Pmd5cb->saveDigestAsFile(md5filename);
	if ( Pindex )
		Pindex->flush(std::string(indexfilename));

	return EXIT_SUCCESS;
}

int main(int argc, char * argv[])
{
	try
	{
		libmaus2::util::ArgInfo const arginfo(argc,argv);

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if (
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if (
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;

				std::vector< std::pair<std::string,std::string> > V;

				V.push_back ( std::pair<std::string,std::string> ( "verbose=<["+::biobambam2::Licensing::formatNumber(getDefaultVerbose())+"]>", "print progress report" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("inputformat=<[")+getDefaultInputFormat()+"]>", std::string("input format (") + libmaus2::bambam::BamMultiAlignmentDecoderFactory::getValidInputFormats() + ")" ) );
				V.push_back ( std::pair<std::string,std::string> ( "I=<[stdin]>", "input filename (standard input if unset)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "inputthreads=<[1]>", "input helper threads (for inputformat=bam only, default: 1)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "outputthreads=<[1]>", "input helper threads (for inputformat=bam only, default: 1)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "numthreads=<[1]>", "number of worker threads (default: 1)" ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
				return EXIT_SUCCESS;
			}

		return bamfiltermc(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
